### PR TITLE
Add Connectionist Temporal Classification loss function.

### PR DESCRIPTION
**Pull request by jnishi**
I added Connectionist Temporal Classification(CTC). It is a loss function of sequence labeling where where the alignment between the inputs and target is unknown.

### CODE REVIEW

**Comment by jnishi** [comment]
This is still work in progress. I implements forward functions, but backward function is not finished yet.

**Comment by jnishi** [comment]
I implemented backward functions. Please check this.

**Comment by jnishi** [comment]
Sorry, this does not supported minibatch. I'll support it.

**Comment by jheymann85** [comment]
Nice work. But to make it usable for real-world cases (e.g. longer sequences), you should calculate the probabilities in the log domain. Otherwise you hit numerical limits pretty fast.

**Comment by jnishi** [comment]
Thanks for comments. I also need to calculate in the log domain with my sample. I'll support it.

**Comment by jnishi** [comment]
Merged {forward, backward}_{cpu, gpu}. (Mini-batch is not supported yet.)

**Comment by jnishi** [comment]
Supported min-batch, cupy, and computation in log-domain. Please check.

File: chainer/functions/loss/ctc.py [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+                labels_prob[c] = self.logsumexp(xp.take(multiply, pos))
+        else:
+            cuda.cupy.ElementwiseKernel(
+                'raw float32 x, raw I y, I s',
+                'float32 z',
+                '''
+                float value = z;
+                for(int index=0;index<s;++index){
+                    if(y[index] == i){
+                        if(value > x[index]){
```
Comment (okuta) [inline chainer/functions/loss/ctc.py]: Index access is high cost in cuda kernel.
It is better to be assigned to a temporary variable.

File: chainer/functions/loss/ctc.py [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+    def calc_path_probability(self, label_prob, total_prob):
+        xp = cuda.get_array_module(label_prob)
+        return xp.exp(label_prob - total_prob)
+
+    def calc_totalprob(self, forward_prob, backward_prob):
+        return self.logsumexp(forward_prob[0] + backward_prob[0])
+
+    def softmax(self, x, axis=None):
+        xp = cuda.get_array_module(x)
+        val = x - xp.amax(x, axis=axis).reshape(x.shape[0], 1)
```
Comment (okuta) [inline chainer/functions/loss/ctc.py]: it is better to use `keepdims` options in amax.

File: chainer/functions/loss/ctc.py [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+
+        for i in range(2, len(in_types)):
+            x_type = in_types[1]
+            utils.type_check.expect(
+                x_type.dtype == numpy.float32,
+                x_type.shape == x_basetype.shape,
+            )
+
+    def log_matrix(self, x):
+        xp = cuda.get_array_module(x)
```
Comment (okuta) [inline chainer/functions/loss/ctc.py]: `cuda.get_array_module` call has overhead.
It is better to make `xp` as an argument of the function.

File: chainer/functions/loss/ctc.py [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+
+            # calc backward probability
+            y_inv = yseq[len(yseq) - t - 1]
+            backward_prob = self.log_dot(backward_prob, rr)
+            beta += backward_prob[::-1],
+            backward_prob = xp.take(y_inv, path[::-1]) + backward_prob
+        return alpha, beta[::-1]
+
+    def convert_inputs(self, labels, yseq, index):
+        log_yseq = ()
```
Comment (okuta) [inline chainer/functions/loss/ctc.py]: `log_yseq = [self.log_matrix(y[index]) for y in yseq]` is better.

File: chainer/functions/loss/ctc.py [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+        self.zero_padding = -10000000000.0
+
+    def check_type_forward(self, in_types):
+        utils.type_check.expect(in_types.size() > 1)
+        l_type = in_types[0]
+        utils.type_check.expect(l_type.dtype == numpy.int32)
+
+        x_basetype = in_types[1]
+
+        for i in range(2, len(in_types)):
```
Comment (okuta) [inline chainer/functions/loss/ctc.py]: Please use `six` for python 2/3 compatibility.

File: chainer/functions/loss/ctc.py [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+    def logsumexp(self, a):
+        xp = cuda.get_array_module(a)
+        vmax = xp.amax(a)
+        res = xp.log(xp.sum(xp.exp(a - vmax)))
+        return (res + vmax).astype(numpy.float32)
+
+    # path probablity to label probability
+    def label_probability(self, label_size, path, multiply):
+        xp = cuda.get_array_module(path)
+        chars = set([c for c in path])
```
Comment (okuta) [inline chainer/functions/loss/ctc.py]: `chars = {c for c in path}` is better.

File: chainer/functions/loss/ctc.py [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+
+    def log_dot(self, prob, rr):
+        xp = cuda.get_array_module(prob)
+        res = xp.zeros(prob.shape)
+        rtrans = xp.swapaxes(rr, 1, 0)
+        for i in range(rtrans.shape[0]):
+            res[i] = self.logsumexp(prob + rtrans[i])
+        return res
+
+    def logsumexp(self, a):
```
Comment (okuta) [inline chainer/functions/loss/ctc.py]: Module level definition is better because this function don't use `self`.

File: chainer/functions/loss/ctc.py [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+                'reduce_probability')(multiply.astype(numpy.float32),
+                                      path, path.shape[0], labels_prob)
+        return labels_prob
+
+    def calc_trans(self, path, yseq, rr):
+        xp = cuda.get_array_module(yseq[0])
+        forward_prob = self.log_matrix(xp.eye(path.shape[0])[0])
+        backward_prob = self.log_matrix(xp.eye(path.shape[0])[0])
+
+        alpha = ()
```
Comment (okuta) [inline chainer/functions/loss/ctc.py]: List type is better?

File: chainer/functions/loss/ctc.py [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+                       self.blank_symbol, dtype=int)
+        for i in range(label_length):
+            path[i * 2 + 1] = labels[i]
+        return path
+
+    def log_dot(self, prob, rr):
+        xp = cuda.get_array_module(prob)
+        res = xp.zeros(prob.shape)
+        rtrans = xp.swapaxes(rr, 1, 0)
+        for i in range(rtrans.shape[0]):
```
Comment (okuta) [inline chainer/functions/loss/ctc.py]: I think that the loop is not required by using the `axis` in `logsumexp`.

File: chainer/functions/loss/ctc.py [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+        return self.logsumexp(forward_prob[0] + backward_prob[0])
+
+    def softmax(self, x, axis=None):
+        xp = cuda.get_array_module(x)
+        val = x - xp.amax(x, axis=axis).reshape(x.shape[0], 1)
+        val = xp.exp(val)
+        return val / xp.sum(val, axis=axis).reshape(x.shape[0], 1)
+
+    def activate(self, yseq):
+        result = ()
```
Comment (okuta) [inline chainer/functions/loss/ctc.py]: `return [self.softmax(y, axis=1) for y in yseq]` is better.

File: chainer/functions/loss/ctc.py [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+        return labels[index], log_yseq
+
+    def calc_path_probability(self, label_prob, total_prob):
+        xp = cuda.get_array_module(label_prob)
+        return xp.exp(label_prob - total_prob)
+
+    def calc_totalprob(self, forward_prob, backward_prob):
+        return self.logsumexp(forward_prob[0] + backward_prob[0])
+
+    def softmax(self, x, axis=None):
```
Comment (okuta) [inline chainer/functions/loss/ctc.py]: `axis` option always `1`?

File: chainer/functions/loss/ctc.py [hunk 12]
```diff
@@ -0,0 +1,247 @@
...
+            forward_prob_trans, backward_prob_trans\
+                = self.calc_trans(path, y, rr)
+            loss += - self.logsumexp(forward_prob_trans[-1]
+                                     + backward_prob_trans[-1])
+        loss /= batch_size
+        return utils.force_array(loss).astype(numpy.float32),
+
+    def backward(self, inputs, grad_output):
+        yseq = self.activate(inputs[1::])
+        delta = []
```
Comment (okuta) [inline chainer/functions/loss/ctc.py]: `delta = list(yseq)` is better.

**Comment by jnishi** [comment]
Thanks for comments. I fixed comments you pointed. Please check them again.

**Comment by jnishi** [comment]
Fixes order of arguments, and checks size of input data of dimensions. Please check.

File: chainer/functions/__init__.py [hunk 13]
```diff
@@ -47,6 +48,9 @@
 
 ClippedReLU = clipped_relu.ClippedReLU
 clipped_relu = clipped_relu.clipped_relu
+connectionist_temporal_classification \
+    = ctc.connectionist_temporal_classification
+ConnectionistTemporalClassification = ctc.connectionist_temporal_classification
```
Comment (unnonouno) [inline chainer/functions/__init__.py]: Is this line true?

File: chainer/functions/loss/ctc.py [hunk 14]
```diff
@@ -0,0 +1,235 @@
+from chainer import cuda
+from chainer import function
+from chainer import utils
+import numpy
+import six
+
```
Comment (unnonouno) [inline chainer/functions/loss/ctc.py]: Change import order like this.

```
import numpy
import six

from chainer import cuda
from chainer import function
from chainer import utils
```

File: chainer/functions/loss/ctc.py [hunk 15]
```diff
@@ -0,0 +1,235 @@
...
+                        multiply[b, path[b] == c], numpy)
+        else:
+            cuda.cupy.ElementwiseKernel(
+                'raw T x, raw I y, I b_max, I c_max',
+                'T z',
+                '''
+                T value = z;
+                I c = i % b_max, b = i / b_max;
+                int ind[2] = {b, -1};
+                for(int index=0;index<c_max;++index){
```
Comment (unnonouno) [inline chainer/functions/loss/ctc.py]: Insert spaces.

File: tests/chainer_tests/functions_tests/loss_tests/test_ctc.py [hunk 16]
```diff
@@ -0,0 +1,109 @@
+import math
+import numpy
+import unittest
```
Comment (unnonouno) [inline tests/chainer_tests/functions_tests/loss_tests/test_ctc.py]:
```
import math
import unittest

import numpy
```

File: chainer/functions/loss/ctc.py [hunk 17]
```diff
@@ -0,0 +1,235 @@
...
+    values of CTC loss is often overflows. This is avoided by computing
+    backward values before the activation function is applied.
+    '''
+
+    def __init__(self, blank_symbol):
+        self.blank_symbol = blank_symbol
+        self.zero_padding = -10000000000.0
+
+    def check_type_forward(self, in_types):
+        utils.type_check.expect(in_types.size() > 1)
```
Comment (unnonouno) [inline chainer/functions/loss/ctc.py]: I recommend `from chainer.utils import type_check`.

File: chainer/functions/loss/ctc.py [hunk 18]
```diff
@@ -0,0 +1,235 @@
...
+        l_type = in_types[0]
+        utils.type_check.expect(l_type.dtype == numpy.int32)
+
+        x_basetype = in_types[1]
+
+        for i in six.moves.range(2, len(in_types)):
+            x_type = in_types[1]
```
Comment (unnonouno) [inline chainer/functions/loss/ctc.py]: ```in_types[i]```?

File: chainer/functions/loss/ctc.py [hunk 19]
```diff
@@ -0,0 +1,237 @@
...
+    return _logsumexp(prob + xp.swapaxes(rr, 1, 2), xp, axis=2)
+
+
+def _activate(yseq, xp):
+    return [_softmax(y, xp) for y in yseq]
+
+
+class ConnectionistTemporalClassification(function.Function):
+
+    '''To make it usable for real-world cases, this class has two policies below.
```
Comment (unnonouno) [inline chainer/functions/loss/ctc.py]: The first line is brief description in docstrings.

File: chainer/functions/loss/ctc.py [hunk 20]
```diff
@@ -0,0 +1,237 @@
...
+                        multiply[b, path[b] == c], numpy)
+        else:
+            cuda.cupy.ElementwiseKernel(
+                'raw T x, raw I y, I b_max, I c_max',
+                'T z',
+                '''
+                T value = z;
+                I c = i % b_max, b = i / b_max;
+                int ind[2] = {b, -1};
+                for(int index=0; index<c_max; ++index){
```
Comment (unnonouno) [inline chainer/functions/loss/ctc.py]: Fix code style in the kernel. I mean insert spaces beside all operators.

**Comment by jnishi** [comment]
Thanks for comments. I fixed them. Please check.

**Comment by okuta** [comment]
LGTM!
